### PR TITLE
[lint] Applies remaining ruff linter changes to hail/ folder

### PR DESCRIPTION
**Pull request by iris-garden**
This change applies each currently-ignored `ruff` rule progressively; each commit applies one rule.

The changes were applied manually to avoid known issues with the automatic fixes; for example, given the code
```python
return (
    is_container(t)
    or isinstance(t, tstruct)
    or isinstance(t, tunion)
    or isinstance(t, ttuple)
    or isinstance(t, tndarray)
)
```
the automatic fixes produce
```python
return isinstance(t, (tndarray, tstruct, ttuple, tunion))
```
instead of 
```python
return is_container(t) or isinstance(t, (tstruct, tunion, ttuple, tndarray))
```
where not only has the call to `is_container` been removed, but also the order of the `isinstance` comparisons has been changed, which has the potential to produce side effects (though in this case, I don’t think it does).

Similarly, when eliminating assignments to unused variables, I left the right-hand side of the assignment intact in case of side effects, except where I myself wrote the code in question and know there are no side effects produced by it.

See also https://github.com/hail-is/hail/pull/14150 and https://github.com/hail-is/hail/pull/14159.

### CODE REVIEW

File: hail/python/test/hail/expr/test_ndarrays.py [hunk 1]
```diff
@@ -892,7 +892,7 @@ def assert_evals_to_same_svd(nd_expr, np_array, full_matrices=True, compute_uv=T
             assert h.shape == n.shape
 
         k = min(np_array.shape)
-        rank = np.linalg.matrix_rank(np_array)
+        np.linalg.matrix_rank(np_array)
```
Comment (daniel-goldstein) [inline hail/python/test/hail/expr/test_ndarrays.py:895]: This should perhaps be asserted somewhere. Maybe ask patrick?

File: hail/python/test/hail/methods/test_impex.py [hunk 2]
```diff
@@ -1091,7 +1091,7 @@ def test_export_plink_non_default_arguments(self):
     def test_export_plink_quantitative_phenotype(self):
         ds = get_dataset()
         fam_mapping = {'f0': 'fam_id', 'f1': 'ind_id', 'f2': 'pat_id', 'f3': 'mat_id', 'f4': 'is_female', 'f5': 'pheno'}
-        bim_mapping = {'f0': 'contig', 'f1': 'varid', 'f2': 'cm_position', 'f3': 'position', 'f4': 'a1', 'f5': 'a2'}
+        {'f0': 'contig', 'f1': 'varid', 'f2': 'cm_position', 'f3': 'position', 'f4': 'a1', 'f5': 'a2'}
```
Comment (daniel-goldstein) [inline hail/python/test/hail/methods/test_impex.py:1094]: Should also be asserted or deleted

File: hail/python/test/hail/methods/test_impex.py [hunk 3]
```diff
@@ -1150,9 +1150,7 @@ def test_contig_recoding_defaults(self):
             reference_genome='GRCh38',
         )
 
-        rg_random = hl.ReferenceGenome(
-            "random", ['1', '23', '24', '25', '26'], {'1': 10, '23': 10, '24': 10, '25': 10, '26': 10}
-        )
+        hl.ReferenceGenome("random", ['1', '23', '24', '25', '26'], {'1': 10, '23': 10, '24': 10, '25': 10, '26': 10})
```
Comment (daniel-goldstein) [inline hail/python/test/hail/methods/test_impex.py:1153]: This is necessary, but I quite dislike that we made the `ReferenceGenome` constructor side effect. :/

File: hail/python/test/hail/methods/test_impex.py [hunk 4]
```diff
@@ -1285,7 +1283,7 @@ def test_parallel_import(self):
         self.assertEqual(mt.count(), (16, 10))
 
     def test_import_bgen_dosage_and_gp_dosage_function_agree(self):
-        recoding = {'0{}'.format(i): str(i) for i in range(1, 10)}
+        {'0{}'.format(i): str(i) for i in range(1, 10)}
```
Comment (daniel-goldstein) [inline hail/python/test/hail/methods/test_impex.py:1286]: Should be asserted or deleted

File: hail/python/test/hail/methods/test_impex.py [hunk 5]
```diff
@@ -2109,7 +2099,7 @@ def test_import_matrix_table_round_trip(missing, delimiter, header, entry_fun):
     actual = actual.rename({'col_id': 'col_idx'})
 
     row_key = mt.row_key
-    col_key = mt.col_key
+    mt.col_key
```
Comment (daniel-goldstein) [inline hail/python/test/hail/methods/test_impex.py:2102]: Should be asserted or deleted

File: hail/python/test/hail/methods/test_pca.py [hunk 6]
```diff
@@ -176,7 +176,7 @@ def test_blanczos_flags(compute_loadings, compute_scores, transpose):
     # compare absolute values to account for +-1 indeterminacy factor in singular vectors
     U = np.abs(U[:, :k])
     V = np.abs(V[:, :k])
-    Usigma = U * sigma[:k]
+    U * sigma[:k]
```
Comment (daniel-goldstein) [inline hail/python/test/hail/methods/test_pca.py:179]: Should be asserted or deleted

File: hail/python/test/hail/methods/test_qc.py [hunk 7]
```diff
@@ -330,7 +330,7 @@ def test_summarize_variants_ti_tv(self):
 
     def test_charr(self):
         mt = hl.import_vcf(resource('sample.vcf'))
-        es = mt.select_rows().entries()
+        mt.select_rows().entries()
```
Comment (daniel-goldstein) [inline hail/python/test/hail/methods/test_qc.py:333]: Should be asserted or deleted

File: hail/python/test/hail/methods/test_statgen.py [hunk 8]
```diff
@@ -94,7 +94,7 @@ def test_linreg_basic(self):
 
     def test_linreg_pass_through(self):
         phenos = hl.import_table(resource('regressionLinear.pheno'), types={'Pheno': hl.tfloat64}, key='Sample')
-        covs = hl.import_table(
+        hl.import_table(
```
Comment (daniel-goldstein) [inline hail/python/test/hail/methods/test_statgen.py:97]: Should be asserted or deleted

File: hail/python/test/hail/methods/test_statgen.py [hunk 9]
```diff
@@ -561,7 +561,7 @@ def test_weighted_linear_regression(self):
             resource('regressionLinear.pheno'), key='Sample', missing='0', types={'Pheno': hl.tfloat}
         )
 
-        weights = hl.import_table(
+        hl.import_table(
```
Comment (daniel-goldstein) [inline hail/python/test/hail/methods/test_statgen.py:564]: Should be asserted or deleted

File: hail/python/test/hail/table/test_table.py [hunk 10]
```diff
@@ -2516,15 +2516,15 @@ def test_table_randomness_matrix_cols_table():
 
 
 def test_table_randomness_parallelize_with_body_randomness():
-    rt = hl.utils.range_table(20, 3)
+    hl.utils.range_table(20, 3)
```
Comment (daniel-goldstein) [inline hail/python/test/hail/table/test_table.py:2519]: This can be deleted

File: hail/python/test/hail/table/test_table.py [hunk 11]
```diff
@@ -2516,15 +2516,15 @@ def test_table_randomness_matrix_cols_table():
...
     t = hl.Table.parallelize(hl.array([1, 2, 3]).map(lambda x: hl.struct(x=x, r=hl.rand_int64())))
     assert_contains_node(t, ir.TableParallelize)
     t._force_count()  # test with no consumer randomness
     assert_unique_uids(t)
 
 
 def test_table_randomness_parallelize_without_body_randomness():
-    rt = hl.utils.range_table(20, 3)
+    hl.utils.range_table(20, 3)
```
Comment (daniel-goldstein) [inline hail/python/test/hail/table/test_table.py:2527]: This can also be deleted

File: hail/python/test/hail/test_ir.py [hunk 12]
```diff
@@ -46,16 +46,16 @@ def value_irs(self):
         a = ir.Ref('a', env['a'])
         st = ir.Ref('st', env['st'])
         whitenStream = ir.Ref('whitenStream')
-        mat = ir.Ref('mat')
-        aa = ir.Ref('aa', env['aa'])
+        ir.Ref('mat')
+        ir.Ref('aa', env['aa'])
```
Comment (daniel-goldstein) [inline hail/python/test/hail/test_ir.py:50]: These shouldn't be side effecting. I'd add their references to the `value_irs` list below to exercise them.

File: hail/python/test/hail/vds/test_vds_functions.py [hunk 13]
```diff
@@ -28,7 +28,7 @@ def test_lgt_to_gt():
 
 def test_lgt_to_gt_invalid():
     c1 = hl.call(1, 1)
-    c2 = hl.call(1, 1, phased=True)
+    hl.call(1, 1, phased=True)
```
Comment (daniel-goldstein) [inline hail/python/test/hail/vds/test_vds_functions.py:31]: The `c2` variable is referenced in the below commented out assertion. Let's turn that comment into an actual test.

File: hail/python/test/hailtop/utils/test_gcs_requester_pays.py [hunk 14]
```diff
@@ -10,7 +10,6 @@ def test_get_fs_by_requester_pays_config(cls):
     config_1 = "foo"
     config_2 = ("foo", ["bar", "baz", "bat"])
     kwargs_1 = {"gcs_requester_pays_configuration": config_1}
-    kwargs_2 = {"gcs_requester_pays_configuration": config_2}
```
Comment (daniel-goldstein) [inline hail/python/test/hailtop/utils/test_gcs_requester_pays.py:13]: This is still a useful config test case, let's add an assert for it.

**Review by daniel-goldstein — CHANGES_REQUESTED** [review]
This is awesome! Helping to catch so many issues in our tests. My comments are basically places where we should add additional testing or deleting dead lines.

File: hail/python/test/hailtop/utils/test_gcs_requester_pays.py (hunk 14 above)
Comment (iris-garden) [reply hail/python/test/hailtop/utils/test_gcs_requester_pays.py:13]: lines 16-19 test that `fs[config_2]._gcs_kwargs` correctly match up with the individual members of `config_2`; is there something else to test with `kwargs_2` here that i'm missing?

Comment (daniel-goldstein) [reply hail/python/test/hailtop/utils/test_gcs_requester_pays.py:13]: nope! I just missed those, thanks for pointing that out.

**Review by daniel-goldstein — APPROVED** [review]

**Pull request merged**